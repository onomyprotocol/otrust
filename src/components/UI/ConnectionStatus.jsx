import styled from 'styled-components';

export const ConnectionStatus = styled.div`
  display: flex;
  align-items: center;
  gap: 8px;

  color: ${props => props.theme.colors.highlightGreen};
  font-weight: 500;

  &:before {
    content: '';
    display: block;

    width: 12px;
    height: 12px;

    background-color: currentColor;
    border-radius: 50%;
  }
`;
