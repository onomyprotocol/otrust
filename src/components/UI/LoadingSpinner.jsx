import React from 'react';

import './loadingBar.css';

export default function LoadingSpinner() {
  return (
    <div className="loadingSpinner">
      <svg>
        <circle cx="20" cy="20" r="20"></circle>
      </svg>
    </div>
  );
}
