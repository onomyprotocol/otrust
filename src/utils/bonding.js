const a = 100000000;

export function priceAtSupply(supply) {
  return (supply / a) ** 2;
}

// Bonding curve price at specified supply
export function supplyAtPrice(price) {
  //#NOM Sold = sqrt(ETH/NOM) * a
  return Math.sqrt(price) * a;
}

// NOM supply range to ETH
export function NOMsupplyETH(supplyTop, supplyBot) {
  // Integrate over curve to get amount of ETH needed to buy amount of NOM
  // ETH = a/3((supplyNOM_Top/a)^3 - (supplyNOM_Bot/a)^3)
  return (a / 3) * ((supplyTop / a) ** 3 - (supplyBot / a) ** 3);
}

// Returns Buy Quote for the purchase of NOM based on amount of ETH
// Parameters:
// Input
// amount: amount of ETH
// Output
// output: amount of NOM
export function ETHtoNOM(amount, supply) {
  const price = priceAtSupply(supply);
  // 1. Determine supply bottom
  // Price bottom is 1% above priceBondCurve
  const priceBot = price * 1.01;
  const supplyBot = supplyAtPrice(priceBot);
  // 2. Integrate over curve, and solve for supply top
  // (3*ETH/a + (supplyNOM_Bot/a)^3)^(1/3)
  const supplyTop = a * ((3 * amount) / a + supplyBot ** 3 / a ** 3) ** (1 / 3);
  // 3. Subtract supply bottom from top to get amount NOM for amount ETH
  const diff = supplyTop - supplyBot;
  return { supplyBot, supplyTop, diff };
}

// Returns Sell Quote: NOM for ETH (Dec 18)
// Parameters:
// Input
// uint256 amountNOM: amount of NOM to be sold (18 decimal)
// Output
// uint256: amount of ETH paid in Wei or ETH (18 decimal)
export function NOMtoETH(amount, supplyTop) {
  const supplyBot = supplyTop - amount;
  const diff = 0.99 * NOMsupplyETH(supplyTop, supplyBot);
  return { supplyBot, supplyTop, diff };
}
