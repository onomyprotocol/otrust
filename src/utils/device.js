const size = {
  mobileS: '320px',
  mobileM: '375px',
  mobileL: '425px',
  tablet: '768px',
  laptop: '1024px',
  desktopS: '1366px',
  desktopM: '1920px',
  desktopL: '2560px',
};

export const device = {
  mobileS: `(min-width: ${size.mobileS})`,
  mobileM: `(min-width: ${size.mobileM})`,
  mobileL: `(min-width: ${size.mobileL})`,
  tablet: `(min-width: ${size.tablet})`,
  laptop: `(min-width: ${size.laptop})`,
  desktopS: `(min-width: ${size.desktopS})`,
  desktopM: `(min-width: ${size.desktopM})`,
  desktopL: `(min-width: ${size.desktopL})`,
};
