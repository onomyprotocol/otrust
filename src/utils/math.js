import { BigNumber } from 'bignumber.js';

export function chopFloat(value, precision) {
  return Math.round(value * 10 ** precision) / 10 ** precision;
}

export function randomInt(min, max) {
  return Math.floor(Math.random() * (max - min + 1) + min);
}

export function truncate(str, maxDecimalDigits) {
  if (str.includes('.')) {
    const parts = str.split('.');
    return parts[0] + '.' + parts[1].slice(0, maxDecimalDigits);
  }
  return str;
}

export function format18(bignumber) {
  return bignumber.div(new BigNumber(10 ** 18));
}

export function parse18(bignumber) {
  return bignumber.times(new BigNumber(10 ** 18));
}

export var isNumber = function isNumber(value) {
  return typeof value === 'number' && isFinite(value);
};
