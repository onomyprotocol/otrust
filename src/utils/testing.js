import { ThemeProvider } from 'styled-components';
import ApolloClient, { InMemoryCache } from 'apollo-boost';
import { ApolloProvider } from '@apollo/client';
import { render } from '@testing-library/react';
import { BigNumber } from 'bignumber.js';

import { darkNew } from 'Theme/theme';
import { ChainContext } from '../context/chain/ChainContext';
import { ExchangeContext } from '../context/exchange/ExchangeContext';

const client = new ApolloClient({
  uri: process.env.REACT_APP_GRAPHQL_ENDPOINT,
  cache: new InMemoryCache(),
});

export const renderWithTheme = (Component, props, children) => {
  if (children) {
    return render(
      <ThemeProvider theme={darkNew}>
        <Component {...props}>{children}</Component>
      </ThemeProvider>,
    );
  }
  return render(
    <ThemeProvider theme={darkNew}>
      <Component {...props} />
    </ThemeProvider>,
  );
};

export const renderWithContext = (Component, props) => {
  return render(
    <ThemeProvider theme={darkNew}>
      <ApolloProvider client={client}>
        <ChainContext.Provider value={{ supplyNOM: new BigNumber(0), theme: darkNew }}>
          <ExchangeContext.Provider value={{ askAmount: 0, bidAmount: 0, bidDenom: 'strong' }}>
            <Component {...props} />
          </ExchangeContext.Provider>
        </ChainContext.Provider>
      </ApolloProvider>
    </ThemeProvider>,
  );
};
