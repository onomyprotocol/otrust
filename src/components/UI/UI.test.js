import { cleanup } from '@testing-library/react';

import LoadingSpinner from './LoadingSpinner';
import { AccentButton, Button, InvisibleButton } from './Button';
import { Container, FullBackgroundContainer } from './Container';
import { Panel } from './Panel';
import { ConnectionStatus } from './ConnectionStatus';
import { renderWithTheme } from '../../utils/testing';

const testProps = {
  width: 50,
};

describe('Given the AccentButton component and width is provided in the props', () => {
  describe('when the component is rendered', () => {
    afterEach(cleanup);

    it('should match the snapshot', () => {
      const { asFragment } = renderWithTheme(AccentButton, testProps);
      expect(asFragment()).toMatchSnapshot();
    });
  });
});

describe('Given the AccentButton component and width is not provided in the props', () => {
  describe('when the component is rendered', () => {
    afterEach(cleanup);

    it('should match the snapshot', () => {
      const { asFragment } = renderWithTheme(AccentButton);
      expect(asFragment()).toMatchSnapshot();
    });
  });
});

describe('Given the Button component and width is provided in the props', () => {
  describe('when the component is rendered', () => {
    afterEach(cleanup);

    it('should match the snapshot', () => {
      const { asFragment } = renderWithTheme(Button, testProps);
      expect(asFragment()).toMatchSnapshot();
    });
  });
});

describe('Given the Button component and width is not provided in the props', () => {
  describe('when the component is rendered', () => {
    afterEach(cleanup);

    it('should match the snapshot', () => {
      const { asFragment } = renderWithTheme(Button);
      expect(asFragment()).toMatchSnapshot();
    });
  });
});

describe('Given the InvisibleButton component', () => {
  describe('when the component is rendered', () => {
    afterEach(cleanup);

    it('should match the snapshot', () => {
      const { asFragment } = renderWithTheme(InvisibleButton);
      expect(asFragment()).toMatchSnapshot();
    });
  });
});

describe('Given the Container component', () => {
  describe('when the component is rendered', () => {
    afterEach(cleanup);

    it('should match the snapshot', () => {
      const { asFragment } = renderWithTheme(Container, undefined, 'Children');
      expect(asFragment()).toMatchSnapshot();
    });
  });
});

describe('Given the FullBackgroundContainer component', () => {
  describe('when the component is rendered', () => {
    afterEach(cleanup);

    it('should match the snapshot', () => {
      const { asFragment } = renderWithTheme(FullBackgroundContainer, 'img', 'Children');
      expect(asFragment()).toMatchSnapshot();
    });
  });
});

describe('Given the LoadingSpinner component', () => {
  describe('when the component is rendered', () => {
    afterEach(cleanup);

    it('should match the snapshot', () => {
      const { asFragment } = renderWithTheme(LoadingSpinner);
      expect(asFragment()).toMatchSnapshot();
    });
  });
});

describe('Given the ConnectionStatus component', () => {
  describe('when the component is rendered', () => {
    afterEach(cleanup);

    it('should match the snapshot', () => {
      const { asFragment } = renderWithTheme(ConnectionStatus);
      expect(asFragment()).toMatchSnapshot();
    });
  });
});

describe('Given the Panel component', () => {
  describe('when the component is rendered', () => {
    afterEach(cleanup);

    it('should match the snapshot', () => {
      const { asFragment } = renderWithTheme(Panel);
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
