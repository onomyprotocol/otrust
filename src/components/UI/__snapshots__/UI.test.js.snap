// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the AccentButton component and width is not provided in the props when the component is rendered should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="sc-gtsrHT cyZncv"
  />
</DocumentFragment>
`;

exports[`Given the AccentButton component and width is provided in the props when the component is rendered should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="sc-gtsrHT kpLgTl"
    width="50"
  />
</DocumentFragment>
`;

exports[`Given the Button component and width is not provided in the props when the component is rendered should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="sc-bdnxRM gLKfsM"
  />
</DocumentFragment>
`;

exports[`Given the Button component and width is provided in the props when the component is rendered should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="sc-bdnxRM fPQjMu"
    width="50"
  />
</DocumentFragment>
`;

exports[`Given the ConnectionStatus component when the component is rendered should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-iCoGMd kkpqdJ"
  />
</DocumentFragment>
`;

exports[`Given the Container component when the component is rendered should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-eCApnc juWvpr"
  >
    Children
  </div>
</DocumentFragment>
`;

exports[`Given the FullBackgroundContainer component when the component is rendered should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-jSFjdj fylwSA"
  >
    Children
  </div>
</DocumentFragment>
`;

exports[`Given the InvisibleButton component when the component is rendered should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="sc-hKFxyN kTRXmr"
  />
</DocumentFragment>
`;

exports[`Given the LoadingSpinner component when the component is rendered should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="loadingSpinner"
  >
    <svg>
      <circle
        cx="20"
        cy="20"
        r="20"
      />
    </svg>
  </div>
</DocumentFragment>
`;

exports[`Given the Panel component when the component is rendered should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-gKAaRy bAUKSe"
  />
</DocumentFragment>
`;
